utils/sync/plocker_bench_test.go

```diff
@@ -0,0 +1,103 @@
+package sync_test
+
+import (
+	"strconv"
+	gsync "sync"
+	"testing"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/utils/sync"
+	"github.com/stretchr/testify/require"
+)
+
+// Running tool: go test -benchmem -run=^$ -bench ^BenchmarkPartitionLocker$ github.com/rudderlabs/rudder-server/utils/sync -v -count=1
+
+// goos: darwin
+// goarch: arm64
+// pkg: github.com/rudderlabs/rudder-server/utils/sync
+// BenchmarkPartitionLocker
+// BenchmarkPartitionLocker/mutex
+// BenchmarkPartitionLocker/mutex-10         	      20	  57351981 ns/op	  162918 B/op	    2121 allocs/op
+// BenchmarkPartitionLocker/cond
+// BenchmarkPartitionLocker/cond-10          	      20	  56779356 ns/op	  149106 B/op	    2030 allocs/op
```

Comment:
This is what I'm getting on my desktop :thinking: 

```
cpu: 12th Gen Intel(R) Core(TM) i9-12900K
BenchmarkPartitionLocker
BenchmarkPartitionLocker/mutex
BenchmarkPartitionLocker/mutex-24         	      18	  58423921 ns/op
BenchmarkPartitionLocker/cond
BenchmarkPartitionLocker/cond-24          	      14	  75101069 ns/op
PASS
ok  	github.com/rudderlabs/rudder-server/utils/sync	6.337s
```

Reply:
The big delta happens when most of my threads (i.e. 21/24) are busy. If I free most of them up there is less difference but still it gives me cond as slower:

```
cpu: 12th Gen Intel(R) Core(TM) i9-12900K
BenchmarkPartitionLocker
BenchmarkPartitionLocker/mutex
BenchmarkPartitionLocker/mutex-24         	      21	  56724928 ns/op
BenchmarkPartitionLocker/cond
BenchmarkPartitionLocker/cond-24          	      18	  63376837 ns/op
PASS
ok  	github.com/rudderlabs/rudder-server/utils/sync	5.434s
```

With a busy CPU I consistently get a big diff between the 2.

Reply:
Is the difference still comparable? E.g. is it 10-20% slower or is it orders of magnitude slower? In any case, `75ms` for having 1000 goroutines locking & unlocking 20 mutexes and sleeping for `1ms` in between is not an issue, right?

Reply:
With a busy CPU it is consistently ~30% slower :disappointed: 

```
BenchmarkPartitionLocker/mutex-24         	      20	  59212050 ns/op
BenchmarkPartitionLocker/cond-24          	      16	  77213625 ns/op
```

![Screenshot from 2023-01-27 14-50-58](https://user-images.githubusercontent.com/1288256/215103020-eaa5ce46-b03c-4df3-b5b1-dd90a530bbbf.png)

If it's slower and clearly also more complex, why use it?
I like the implementation, it's a smart use of `cond` but I'm not sure I get it.

Reply:
It is more cpu intensive but more memory efficient. It is slower in some architectures & scenarios but faster in others. Complexity is of course relevant, since it mostly boils down to a person's understanding of `sync.Cond` 😄 

I would argue that any of these solutions would do just fine coping with our workloads since they are comparable, but let's vote!

❤️ `sync.Cond`
🚀 map of `sync.Mutex`

Reply:
My take is that most of the time we want our CPU busy doing things and that should be the normal state in a MultiTenant architecture so we can assume that most of the time it's going to be 30% slower. The advantage in terms of memory isn't a lot to begin with i.e. ~%5. 

So to me it boils down to deciding between:
* have a 5% more memory efficient algorithm that is 30% less CPU efficient and that is more complex to understand/maintain/debug
* have a common pattern that consumes 5% more memory but it is 30% faster

> Complexity is of course relevant, since it mostly boils down to a person's understanding of sync.Cond

It is not just that! :smile: Tests are also added and benchmarks and it is more code overall in the codebase.
In your tests you're also creating wrappers around mutexes. Worst case, we might happen to have to debug some code live that is using your version of a mutex and by then the guy on call would have forgot how it works :smile:

Reply:
It depends on the scenario and the architecture I guess since with `numKeys: 100000` and `numGoroutines: 500000` I am getting the following results:

```
BenchmarkPartitionLocker/mutex
BenchmarkPartitionLocker/mutex-10         	       1	1330982500 ns/op	336118704 B/op	 2583238 allocs/op
BenchmarkPartitionLocker/cond
BenchmarkPartitionLocker/cond-10          	       1	1295121292 ns/op	127760256 B/op	 1918228 allocs/op
```
I don't see any of the two solutions as true bottlenecks, but also understand the value of simplicity! Nevertheless, this was a fun exercise!